### ISSUE TITLE
fix: backward graph capture — move gradient clipping outside graph boundary (PMAT-464)

### ISSUE THREAD

**Issue by noahgift** [paiml/qwen-train-canary#23]
## Summary

CUDA graph backward capture is blocked by **per-layer gradient clipping D2H sync** inside the backward loop. `squared_sum_cuda()` calls `stream.synchronize()` — not capturable by CUDA graph.

## Five-Whys Root Cause

1. Backward graph capture fails → `CUDA_ERROR_STREAM_CAPTURE_UNSUPPORTED`
2. Stream capture unsupported → D2H synchronization inside captured region
3. D2H sync → `squared_sum_cuda()` in `clip_gradients()` forces `stream.synchronize()`
4. Per-layer clipping → gradient norm computed per-layer inside backward loop (168 sync points)
5. Can be fixed → move clipping outside graph, compute norms in batch, single sync

## The Fix

**Current** (NOT capturable):
```
for layer in 27..=0:
  backward(layer)
  norm = squared_sum_cuda(grads)  ← stream.synchronize() HERE
  if norm > threshold:
    gradient_clip_cuda(grads, scale)
  optimizer_step(layer)
```

**Fixed** (capturable):
```
# Phase 1: Backward (captured in graph)
graph_begin()
for layer in 27..=0:
  backward(layer)
graph_end()

# Phase 2: Clip (outside graph, single sync)
for layer in 27..=0:
  squared_sum_launch_cuda(grads, &partial[layer])  ← async
stream.synchronize()  ← ONE sync
total_norm = reduce(partial)
for layer in 27..=0:
  gradient_clip_cuda(grads, threshold / total_norm)

# Phase 3: Optimizer (async, no sync)
for layer in 27..=0:
  optimizer_step(layer)
```

## Files to Change

| File | Lines | Change |
|------|-------|--------|
| `src/finetune/instruct_pipeline.rs` | 2287-2367 | Split backward loop into 3 phases |
| `src/transformer/cuda_block.rs` | 3458-3497 | Decouple clip_gradients from backward |
| `src/autograd/cuda_optim.rs` | 395-401 | Add async variant of squared_sum |

## Expected Impact

- Eliminates 168 D2H sync points per backward pass
- Enables backward graph capture (single `cudaGraphLaunch` for all 28 layers)
- At batch=4: ~10-30% throughput improvement (sync overhead vs kernel time)
- Combined with forward graph: full training loop in 2 graph launches

## Contract
`cuda-graph-backward-v1.yaml`

## Falsification
F-GRAPH-BWD-004: Global clipping may differ from per-layer clipping. Extended 1000-step canary needed to verify loss trajectory parity.
